lektor/imagetools.py

```diff
@@ -440,40 +489,74 @@ def process_image(ctx, source_image, dst_filename, width, height=None,
 
 
-def make_thumbnail(ctx, source_image, source_url_path, width, height=None,
-                   crop=False, quality=None):
+def make_thumbnail(ctx, source_image, source_url_path,
+                   width=None, height=None, mode=THUMBNAIL_MODE.DEFAULT,
+                   upscale=None, quality=None):
     """Helper method that can create thumbnails from within the build process
     of an artifact.
     """
+    if width is None and height is None:
+        raise ValueError("Must specify at least one of width or height.")
+
+    if upscale is None:
+        upscale = {
+            THUMBNAIL_MODE.FIT: False,
+            THUMBNAIL_MODE.CROP: True,
+            THUMBNAIL_MODE.STRETCH: True,
+        }[mode]
+
+    # fallback to "fit" in case of erroneous arguments (but preserve
+    # the `upscale` defaults as defined above).
+    if mode != THUMBNAIL_MODE.FIT and (width is None or height is None):
+        warnings.warn(
+            '"%s" mode requires both `width` and `height` to be defined. '
+            'Falling back to "fit" mode.`' % THUMBNAIL_MODE.__RMAPPING[mode].lower()
+        )
+        mode = THUMBNAIL_MODE.FIT
+
     with open(source_image, 'rb') as f:
         format, source_width, source_height = get_image_info(f)
-        if format == 'unknown':
-            raise RuntimeError('Cannot process unknown images')
-
-    suffix = get_suffix(width, height, crop=crop, quality=quality)
-    dst_url_path = get_dependent_url(source_url_path, suffix,
-                                     ext=get_thumbnail_ext(source_image))
-    report_height = height
 
-    if height is None:
-        # we can only crop if a height is specified
-        crop = False
-        report_height = computed_height(source_image, width, source_width,
-                                        source_height)
+    if format is None:
+        raise RuntimeError('Cannot process unknown images')
 
     # If we are dealing with an actual svg image, we do not actually
-    # resize anything, we just return it.  This is not ideal but it's
+    # resize anything, we just return it. This is not ideal but it's
     # better than outright failing.
     if format == 'svg':
         return Thumbnail(source_url_path, width, height)
 
+    if not upscale:
+        def _original():
+            return Thumbnail(source_url_path, source_width, source_height)
+
+        if height is None:
+            if source_width < width:
+                return _original()
+        elif width is None:
+            if source_height < height:
+                return _original()
+        else:
+            if source_width < width and source_height < height:
+                return _original()
+
+    suffix = get_suffix(width, height, mode, quality=quality)
+    dst_url_path = get_dependent_url(source_url_path, suffix,
+                                     ext=get_thumbnail_ext(source_image))
+
+    if mode == THUMBNAIL_MODE.FIT:
+        computed_width, computed_height = compute_dimensions(
```

Comment:
And in... wait for it... the thumbnail-generator plugin. :)
I personally wouldn't rename it, because well, I'm too lazy to come up with something meaningful, and "compute_fit_dimensions" is perfectly meaningless. And anything else doesn't compute anything, it just applies them bluntly. Your choice.

Reply:
I'm having trouble coming up with anything better than "compute_fit_dimensions", which I suppose you're right, doesn't really add much value. It I think of something I'll let you know, but as is is fine for now.

Reply:
Haha, sorry if I'm making your life harder, guys, let me know if I need to change anything in `thumbnail-generator` and i can do it.